File: time_complexity_analyzer/analyzer/InstrumentedPrototype.java

package time_complexity_analyzer.analyzer;

import java.io.PrintWriter;
import java.io.File;
import java.io.IOException;
import java.util.HashMap;

public class InstrumentedPrototype {
    public HashMap<Integer, Long> lineInfoLastStart = new HashMap<>();
    public HashMap<Integer, Long> lineInfoTotal = new HashMap<>();

    private long getLastLineInfo(int lineNumber) {
        if (lineInfoLastStart.containsKey(lineNumber)) {
            return lineInfoLastStart.get(lineNumber);
        } else if (lineNumber > 1) {
            return getLastLineInfo(lineNumber - 1);
        }
        return 0L;
    }

    public void mySortFunction(int[] array) {
        this.lineInfoLastStart.put(2, System.nanoTime());
        for (int i = 0; i < array.length; i++) {
            this.lineInfoTotal.put(2, this.lineInfoTotal.getOrDefault(2, 0L) + System.nanoTime() - getLastLineInfo(2));
            this.lineInfoLastStart.put(3, System.nanoTime());
            for (int j = i + 1; j < array.length; j++) {
                this.lineInfoTotal.put(3,
                        this.lineInfoTotal.getOrDefault(3, 0L) + System.nanoTime() - getLastLineInfo(3));
                this.lineInfoLastStart.put(4, System.nanoTime());
                if (array[i] > array[j]) {
                    this.lineInfoTotal.put(4,
                            this.lineInfoTotal.getOrDefault(4, 0L) + System.nanoTime() - getLastLineInfo(4));
                    this.lineInfoLastStart.put(5, System.nanoTime());
                    int temp = array[i];
                    this.lineInfoTotal.put(5,
                            this.lineInfoTotal.getOrDefault(5, 0L) + System.nanoTime() - getLastLineInfo(5));
                    this.lineInfoLastStart.put(6, System.nanoTime());
                    array[i] = array[j];
                    this.lineInfoTotal.put(6,
                            this.lineInfoTotal.getOrDefault(6, 0L) + System.nanoTime() - getLastLineInfo(6));
                    this.lineInfoLastStart.put(7, System.nanoTime());
                    array[j] = temp;
                    this.lineInfoTotal.put(7,
                            this.lineInfoTotal.getOrDefault(7, 0L) + System.nanoTime() - getLastLineInfo(7));
                }
            }
        }
    }

    public static void main(String[] args) {
        InstrumentedPrototype p = new InstrumentedPrototype();
        p.mySortFunction(new int[] { 3, 4, 5, 6, 0 });
        long startTime = System.nanoTime();
        try (PrintWriter pw = new PrintWriter(new File("time_complexity_analyzer/analyzer/output_java.txt"))) {
            pw.println("Function execution time: " + (System.nanoTime() - startTime) + " ns");
            pw.println(p.lineInfoTotal.toString());
        } catch (IOException ex) {
        }
    }
}

File: time_complexity_analyzer/analyzer/InstrumentedPrototype.cpp

#include <iostream>
#include <fstream>
#include <chrono>
#include <map>
#include <vector>

class InstrumentedPrototype {
public:
    std::map<int, std::chrono::high_resolution_clock::time_point> lineInfoLastStart;
    std::map<int, long long> lineInfoTotal;

    InstrumentedPrototype() {
        // Constructor
    }

    std::chrono::high_resolution_clock::time_point getLastLineInfo(int lineNumber) {
        auto it = lineInfoLastStart.find(lineNumber);
        if (it != lineInfoLastStart.end()) {
            return it->second;
        } else if (lineNumber > 1) {
            return getLastLineInfo(lineNumber - 1);
        }
        return std::chrono::high_resolution_clock::now();
    }
    
void mySortFunction(std::vector<int>& array) {

            lineInfoLastStart[1] = std::chrono::high_resolution_clock::now();
                for (size_t i = 0; i < array.size(); ++i) {
            lineInfoTotal[1] += std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - getLastLineInfo(1)).count();

            lineInfoLastStart[2] = std::chrono::high_resolution_clock::now();
                    for (size_t j = i + 1; j < array.size(); ++j) {
            lineInfoTotal[2] += std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - getLastLineInfo(2)).count();

            lineInfoLastStart[3] = std::chrono::high_resolution_clock::now();
                        if (array[i] > array[j]) {
            lineInfoTotal[3] += std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - getLastLineInfo(3)).count();

            lineInfoLastStart[4] = std::chrono::high_resolution_clock::now();
                            int temp = array[i];
            lineInfoTotal[4] += std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - getLastLineInfo(4)).count();

            lineInfoLastStart[5] = std::chrono::high_resolution_clock::now();
                            array[i] = array[j];
            lineInfoTotal[5] += std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - getLastLineInfo(5)).count();

            lineInfoLastStart[6] = std::chrono::high_resolution_clock::now();
                            array[j] = temp;
            lineInfoTotal[6] += std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now() - getLastLineInfo(6)).count();
            }
        }
    }
    }

    void execute() {
        std::vector<int> array = {3, 4, 5, 6, 0};
        mySortFunction(array);
    }

    void saveResults() {
        std::ofstream outFile("time_complexity_analyzer/analyzer/output_cpp.txt");
        for (auto& pair : lineInfoTotal) {
            outFile << "Line " << pair.first << ": " << pair.second << "ns\n";
        }
        outFile.close();
    }
};

int main() {
    InstrumentedPrototype p;
    p.execute();
    p.saveResults();
    return 0;
}
